File: src/main/kotlin/script/lua/World.kt
package org.lain.engine.script.lua

import org.lain.cyberia.ecs.Component
import org.lain.cyberia.ecs.componentTypeOf
import org.lain.cyberia.ecs.getComponent
import org.lain.cyberia.ecs.setComponent
import org.lain.engine.player.username
import org.lain.engine.script.CoreScriptComponents
import org.lain.engine.script.ScriptComponent
import org.lain.engine.script.ScriptComponentType
import org.lain.engine.util.component.EntityId
import org.lain.engine.world.World
import org.lain.engine.world.invokeCommand
import org.lain.engine.world.setDynamicVoxel
import org.luaj.vm2.Globals
import org.luaj.vm2.LuaTable
import org.luaj.vm2.LuaUserdata
import org.luaj.vm2.LuaValue
import org.luaj.vm2.LuaValue.NIL
import org.luaj.vm2.lib.TwoArgFunction

context(ctx: LuaContext)
fun World.coerceToLua(): LuaUserdata {
    (ctx.worldsList[id.value] as? LuaUserdata) ?.let { return it }
    val userdata = LuaUserdata(this)
    val idValue = luaValue(this@coerceToLua.id.toString())
    val isClient = luaValue(isClient)
    val worldState = worldState.coerceToLua(userdata)
    val meta = object : LuaTable() {
        init {
            set("__index", object : TwoArgFunction() {
                override fun call(self: LuaValue, key: LuaValue): LuaValue {
                    return when (key.tojstring()) {
                        "id" -> idValue
                        "is_client" -> isClient
                        "players" -> LuaTable.tableOf(
                            players.flatMap { listOf(it.username.toLuaValue(), it.coerceToLua()) }.toTypedArray(),
                        )
                        "state" -> worldState
                        else -> ctx.worldTable.get(key)
                    }
                }
            })
        }
    }
    userdata.setmetatable(meta)
    return userdata
}

data class LuaEntityComponent(val table: LuaValue) : Component {
    companion object {
        val TYPE by lazy { componentTypeOf(LuaEntityComponent::class) }
    }
}

context(ctx: LuaContext, world: World)
fun EntityId.coerceToLua(worldV: LuaValue? = null): LuaValue {
    val idValue = luaValue(this)
    val worldValue = worldV ?: world.coerceToLua()
    return getComponent<LuaEntityComponent>()?.table ?: run {
        val t = object : LuaTable() {
            override fun get(key: LuaValue): LuaValue {
                return when (key.tojstring()) {
                    "id" -> idValue
                    "world" -> worldValue
                    else -> ctx.entityTable.get(key)
                }
            }
        }
        setComponent<LuaEntityComponent>(LuaEntityComponent(t))
        t
    }
}

fun LuaValue.coerceToEngineWorld() = this.checkuserdata() as World

context(ctx: LuaContext)
fun Globals.setupEntity() {
    ctx.entityTable.set("_of", twoArgFunction { world, id ->
        val world = world.coerceToEngineWorld()
        with(world) { id.toint().coerceToLua() }
    })
}

context(ctx: LuaContext)
fun Globals.setupWorld() {
    ctx.worldTable.set("_invoke_command", twoArgFunction { self, command ->
        val world = self.coerceToEngineWorld()
        val commandStr = command.tojstring()
        world.invokeCommand(commandStr)
        LuaValue.NIL
    })
    ctx.worldTable.set("_add_entity", oneArgFunction { self ->
        val world = self.coerceToEngineWorld()
        with(world) { world.addEntity().coerceToLua() }
    })
    ctx.worldTable.set("_get_component", threeArgFunction { self, entityId, type ->
        val world = self.coerceToEngineWorld()
        val type = type.coerceToScriptComponentType()
        val entityId = entityId.toint()
        world.getLuaComponent(entityId, type.requireComponent())
    })

    ctx.worldTable.set("_has_component", threeArgFunction { self, entityId, type ->
        val world = self.coerceToEngineWorld()
        val type = type.coerceToScriptComponentType()
        val entityId = entityId.toint()
        luaValue(world.hasLuaComponent(entityId, type.requireComponent()))
    })

    ctx.worldTable.set("_set_component", fourArgFunction { self, entityId, type, component ->
        val world = self.coerceToEngineWorld()
        val type = type.coerceToScriptComponentType()
        val entityId = entityId.toint()
        val component = component.checktable()
        with(world) { entityId.setLuaComponent(component, type.requireComponent()) }
        NIL
    })

    ctx.worldTable.set("_remove_component", threeArgFunction { self, entityId, type ->
        val world = self.coerceToEngineWorld()
        val type = type.coerceToScriptComponentType()
        val entityId = entityId.toint()
        world.removeLuaComponent(entityId, type.requireComponent()) ?: NIL
    })

    ctx.worldTable.set("_get_all_components", twoArgFunction { self, entityId ->
        val world = self.coerceToEngineWorld()
        LuaTable.listOf(
            world.getComponents(entityId.toint())
                .filterIsInstance<ScriptComponent>()
                .filter { it.field is LuaTable }
                .map { it.luaValue }
                .toTypedArray()
        )
    })

    ctx.worldTable.set("_destroy_entity", twoArgFunction { self, entityId ->
        val world = self.coerceToEngineWorld()
        world.destroy(entityId.toint())
        NIL
    })

    ctx.worldTable.set("_iterate", varargsFunction { args ->
        val world = args.arg(1).coerceToEngineWorld()
        val func = args.arg(2).checkfunction()
        val types = setOf(args.arg(3), args.arg(4), args.arg(5), args.arg(6), args.arg(7))
            .filter { !it.isnil() }
        val entityArray = world.componentManager.getComponentArray(LuaEntityComponent.TYPE)

        fun getOrCreateEntityComponent(entityId: EntityId): LuaEntityComponent {
            return entityArray.componentOf(entityId) ?: run {
                with(world) { entityId.coerceToLua() }
                entityArray.componentOf(entityId)!!
            }
        }

        when (types.size) {
            1 -> world.iterate1(types[0].coerceToScriptComponentType().ecsType) { entity, component ->
                val luaEntity = getOrCreateEntityComponent(entity)
                func.invoke(
                    luaEntity.table,
                    component.luaValue
                )
            }

            2 -> world.iterate2(
                types[0].coerceToScriptComponentType().ecsType,
                types[1].coerceToScriptComponentType().ecsType
            ) { entity, component1, component2 ->
                val luaEntity = getOrCreateEntityComponent(entity)
                func.invoke(luaEntity.table, component1.luaValue, component2.luaValue)
            }

            3 -> world.iterate3(
                types[0].coerceToScriptComponentType().ecsType,
                types[1].coerceToScriptComponentType().ecsType,
                types[2].coerceToScriptComponentType().ecsType
            ) { entity, component1, component2, component3 ->
                val luaEntity = getOrCreateEntityComponent(entity)
                func.invoke(
                    arrayOf(luaEntity.table, component1.luaValue, component2.luaValue, component3.luaValue)
                )
            }

            4 -> world.iterate4(
                types[0].coerceToScriptComponentType().ecsType,
                types[1].coerceToScriptComponentType().ecsType,
                types[2].coerceToScriptComponentType().ecsType,
                types[3].coerceToScriptComponentType().ecsType
            ) { entity, component1, component2, component3, component4 ->
                val luaEntity = getOrCreateEntityComponent(entity)
                func.invoke(
                    arrayOf(
                        luaEntity.table,
                        component1.luaValue,
                        component2.luaValue,
                        component3.luaValue,
                        component4.luaValue
                    )
                )
            }

            5 -> world.iterate5(
                types[0].coerceToScriptComponentType().ecsType,
                types[1].coerceToScriptComponentType().ecsType,
                types[2].coerceToScriptComponentType().ecsType,
                types[3].coerceToScriptComponentType().ecsType,
                types[4].coerceToScriptComponentType().ecsType
            ) { entity, component1, component2, component3, component4, component5 ->
                val luaEntity = getOrCreateEntityComponent(entity)
                func.invoke(
                    arrayOf(
                        luaEntity.table,
                        component1.luaValue,
                        component2.luaValue,
                        component3.luaValue,
                        component4.luaValue,
                        component5.luaValue
                    )
                )
            }
        }
        NIL
    })

    ctx.worldTable.set("_set_dynamic_voxel", threeArgFunction { self, pos, networked ->
        val world = self.coerceToEngineWorld()
        val voxelPos = pos.toVoxelPos()
        with(world) {
            val entity = setDynamicVoxel(voxelPos, networked.toboolean())
            entity.setLuaComponent(
                voxelPos.toLuaValue(),
                CoreScriptComponents.DYNAMIC_VOXEL
            )
            entity.coerceToLua()
        }
    })

    ctx.worldTable.set("mark_dirty", threeArgFunction { self, entityId, type ->
        val world = self.coerceToEngineWorld()
        val entity = entityId.toint()
        world.markDirty(entity, type.coerceToScriptComponentType().requireComponent().ecsType)
        LuaValue.NIL
    })
}

val ScriptComponent.luaValue
    get() = field as? LuaValue ?: error("Component not supports lua")

private fun World.hasLuaComponent(entityId: EntityId, componentType: ScriptComponentType): Boolean {
    return hasComponent(entityId, componentType.ecsType)
}

private fun World.getLuaComponent(entityId: EntityId, componentType: ScriptComponentType): LuaValue? {
    val component = getComponent(entityId, componentType.ecsType) ?: return null
    return (component.field as? LuaValue) ?: error("Component ${componentType.id} not supports lua")
}

private fun World.removeLuaComponent(entityId: EntityId, componentType: ScriptComponentType): LuaValue? {
    val component = removeComponent(entityId, componentType.ecsType) ?: return null
    if (component.field !is LuaValue) error("Removed non-lua component with type ${componentType.id}")
    return component.field
}

File: src/test/resources/world.lua
require("core.bridge")
require("core.util")
require("core.component")

--- debug
---@return string
function listworlds()
    return tablestr(worlds)
end

---@type World for EmmyLua
local World = World

---@type Entity for EmmyLua
local Entity = Entity

---@param types ComponentType[]|Component[]
---@param fun fun(entity: number, ...)
function World:iterate(types, fun)
    local mapped_types = map(types, function(elem)
        return elem.type or elem
    end)
    self:__iterate(fun, table.unpack(mapped_types))
end

---@param types ComponentType[]|Component[]
---@param fun fun(world: World, entity: number, ...)
function World:iterate_in(types, fun)
    self:iterate(types, function(entity, ...)
        fun(self, entity, ...)
    end)
end

--------------------------------------------

---@class LocationComponent : Component
---@field vector number[]
LocationComponent = Component.of("core/location")

---@field x number
---@field y number
---@field z number
---@return LocationComponent
function LocationComponent.new(x, y, z)
    assert(x ~= nil, "x must be not null")
    assert(y ~= nil, "y must be not null")
    assert(z ~= nil, "z must be not null")
    return LocationComponent:construct({ vector = { x, y, z } })
end

---@field vector number[]
---@return LocationComponent
function LocationComponent.vector(vector)
    assert(vector ~= nil, "vector must be not null")
    return LocationComponent.new(vector[1], vector[2], vector[3])
end

--------------------------------------------

---@class DynamicVoxelComponent : Component
---@field pos number[]
DynamicVoxelComponent = Component.of("core/voxel/dynamic_voxel")

---@class UseRestrictionComponent : Component
UseRestrictionComponent = Component.of("core/voxel/use_restriction")

---@param voxel_pos number[]
---@param networked boolean default false
---@return Entity
function World:set_dynamic_voxel(voxel_pos, networked)
    assert(voxel_pos, "voxel pos must be not null")
    networked = networked == true
    local entity = self:__set_dynamic_voxel(voxel_pos, networked)
    debug("voxel", "(" .. entity.id .. ") created dynamic voxel, networked = " .. tostring(networked))
    return entity
end

--------------------------------------------

---@param type ComponentType
---@param entity_id number
---@return table
function World:get_component(entity_id, type)
    assert(entity_id, "entity_id must be not null")
    assert(type, "component type must be not null")
    return self:__get_component(entity_id, type.type or type)
end

---@param type ComponentType
---@param entity_id number
---@return Component?
function World:remove_component(entity_id, type)
    assert(entity_id, "entity_id must be not null")
    assert(type, "component type must be not null")
    local component_type = type.type or type
    debug("component", "(" .. entity_id .. ") remove " .. component_type.id)
    return self:__remove_component(entity_id, component_type)
end

---@param entity_id number
---@param component Component?
---@param type ComponentType
function World:set_component(entity_id, component, type)
    assert(entity_id, "entity_id must be not null")
    local component_type = type or component.type
    assert(component_type, "component must be not null")
    debug("component", "(" .. entity_id .. ") set " .. component_type.id)
    return self:__set_component(entity_id, component_type, component)
end

---@param entity_id number
---@param type ComponentType
---@return boolean
function World:has_component(entity_id, type)
    assert(entity_id, "entity_id must be not null")
    local component_type = type.type or type
    return self:__has_component(entity_id, component_type)
end

--------------------------------------------

---@param components Component[]?
---@return Entity
function World:add_entity(components)
    local entity = self:__add_entity()
    if (components ~= nil) then
        for_each(components, function(component) entity:set_component(component) end)
    end
    return entity
end

---@param entity_id number
---@return Component[]
function World:components_of(entity_id)
    assert(entity_id, "entity_id must be not null")
    return self:__get_all_components(entity_id)
end

function World:destroy_entity(entity_id)
    assert(entity_id, "entity_id must be not null")
    debug("entity", entity_id .. " destroyed")
    return self:__destroy_entity(entity_id)
end

--------------------------------------------

---@param world World
---@param id number
---@return Entity
function Entity.of(world, id) return Entity.__of(world, id) end

---@param component Component?
function Entity:set_component(component)
    self.world:set_component(self.id, component)
end

---@param component Component
function Entity:has_component(component)
    return self.world:has_component(self.id, component)
end

---@param component Component
---@return Component?
function Entity:remove_component(component)
    return self.world:remove_component(self.id, component)
end

---@param component Component
---@return Component?
function Entity:get_component(component)
    return self.world:get_component(self.id, component)
end

---@generic T : Component
---@param component T
---@param factory fun(): T
---@return T
function Entity:get_or_create_component(component, factory)
    assert(factory ~= nil, "component factory must be not null")
    assert(component ~= nil, "component type must be not null")
    local component2 = self:get_component(component)
    if (component2 == nil) then
        component2 = factory()
        assert(component ~= nil, "component factory must return not null value")
        self:set_component(component2)
    end
    return component2
end

function Entity:destroy()
    self.world:destroy_entity(self.id)
end

---@return Component[]
function Entity:get_all()
    return self.world:components_of(self.id)
end

---@param component Component
function Entity:mark_dirty(component)
    self.world:mark_dirty(self.id, component.type)
end